### PR TITLE
Improve Snapshot Finalization Ex. Handling

### LINKED ISSUE #49989: [CI] Failure in DedicatedClusterSnapshotRestoreIT.testMasterAndDataShutdownDuringSnapshot
`DedicatedClusterSnapshotRestoreIT.testMasterAndDataShutdownDuringSnapshot` failed in https://gradle-enterprise.elastic.co/s/ollsfmdiyy7ts/console-log/raw with the following:

```
REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.snapshots.DedicatedClusterSnapshotRestoreIT.testMasterAndDataShutdownDuringSnapshot" -Dtests.seed=4090EEDAC757A077 -Dtests.security.manager=true -Dtests.locale=pl-PL -Dtests.timezone=Pacific/Tarawa -Dcompiler.java=12
```

```
org.elasticsearch.snapshots.DedicatedClusterSnapshotRestoreIT > testMasterAndDataShutdownDuringSnapshot FAILED
    org.elasticsearch.snapshots.SnapshotMissingException: [test-repo:test-snap] is missing
```

It seems the problem is here.

```
  1> [2019-12-10T02:00:32,554][WARN ][o.e.s.SnapshotsService   ] [node_tm0] failed to update snapshot state after node removal
  1> [2019-12-10T02:00:32,558][WARN ][o.e.s.SnapshotsService   ] [node_tm0] [test-repo:test-snap/m5-vYhZeSIm67jm7Wz-3ow] failed to finalize snapshot
  1> org.elasticsearch.snapshots.SnapshotException: [test-repo:test-snap/m5-vYhZeSIm67jm7Wz-3ow] failed to update snapshot in repository
  1> 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$finalizeSnapshot$31(BlobStoreRepository.java:812) [main/:?]
  1> 	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [main/:?]
  1> 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository$2.onFailure(BlobStoreRepository.java:1118) [main/:?]
  1> 	at org.elasticsearch.cluster.ClusterStateTaskListener.onNoLongerMaster(ClusterStateTaskListener.java:37) [main/:?]
  1> 	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.onNoLongerMaster(MasterService.java:519) [main/:?]
  1> 	at org.elasticsearch.cluster.service.MasterService$TaskInputs.lambda$onNoLongerMaster$0(MasterService.java:762) [main/:?]
  1> 	at java.util.ArrayList.forEach(ArrayList.java:1540) [?:?]
  1> 	at org.elasticsearch.cluster.service.MasterService$TaskInputs.onNoLongerMaster(MasterService.java:762) [main/:?]
  1> 	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:209) [main/:?]
  1> 	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:151) [main/:?]
  1> 	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [main/:?]
  1> 	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
  1> 	at java.lang.Thread.run(Thread.java:834) [?:?]
  1> Caused by: org.elasticsearch.repositories.RepositoryException: [test-repo] Failed to execute cluster state update [set pending repository generation [test-repo][-1]]
  1> 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository$2.onFailure(BlobStoreRepository.java:1119) ~[main/:?]
  1> 	... 15 more
  1> Caused by: org.elasticsearch.cluster.NotMasterException: no longer master. source: [set pending repository generation [test-repo][-1]]
```

It seems the test does not fully take the fact that we need to form a working cluster for the snapshot to finalise due to #49729 nowadays. I'm looking into that.

### PR DESCRIPTION
**Pull request by original-brownbear**
Like in #49989 we can get into a situation where the setting of
the repository generation (during snapshot finalization) in the cluster
state fails due to master failing over.
In this case we should not try to execute the next cluster state update
that will remove the snapshot from the cluster state. 
Otherwise we may needlessly drop an otherwise fine snapshot from the repository
completely on a master failover via the rare case of the removing of the snapshot from the CS working out because the node that failed the generation update from the blob store repository becoming master just in time to remove the snapshot from the repository.

Note: this won't corrupt the repository, it simply needlessly fails snapshots that should just work out fine on a master failover like the case in #49989

Closes #49989

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -1052,7 +1053,13 @@ protected void doRun() {
             public void onFailure(final Exception e) {
                 Snapshot snapshot = entry.snapshot();
                 logger.warn(() -> new ParameterizedMessage("[{}] failed to finalize snapshot", snapshot), e);
-                removeSnapshotFromClusterState(snapshot, null, e);
+                if (ExceptionsHelper.unwrap(e, NotMasterException.class) != null) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: This feels kind of dirty but IMO it's an ok best-effort work around for now. We simply aren't exposing the internals of the cluster state updates from the `BlobStoreRepository` and I don't see a straight-forward way of doing so with the current `Repository` interface. Also, this is just to help (as in improve the user experience, there's no corruption to be fixed here :)) with the incredibly unlikely corner case showing in the linked test failure were a new master is elected, immediately fails over again and is then finally elected for good.

**Review by tlrx — APPROVED** [review]
LGTM, left a minor comment

**Comment by original-brownbear** [comment]
Thanks Tanguy!

> LGTM, left a minor comment

Where's that comment though :) ?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: You should also handle `FailedToCommitClusterStateException` here (see also TransportMasterNodeAction). Note that I'm not sure how the wrapping comes into play here / whether we need to unwrap.

Comment (tlrx) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: > Note that I'm not sure how the wrapping comes into play here / whether we need to unwrap.

Not sure either, so unwrapping seems safer to me

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Right, added that exception here :)

We gotta unwrap, we're wrapping everything in the CS update callbacks in a `RepositoryException` in `BlobStoreRepository`

**Comment by tlrx** [comment]
> Where's that comment though :) ?

In my browser cache of course :)

> logger.warn(() -> new ParameterizedMessage("[{}] failed to finalize snapshot", snapshot), e);

I think we should only log this if we effectively remove the snapshot from the cluster state and not warn anything if we let the next master finalizes the snapshot.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: `RepositoryException` is a not a `WrapperException`, so the unwrapping won't work. I'm confused how this fixes the test.

**Comment by original-brownbear** [comment]
> I think we should only log this if we effectively remove the snapshot from the cluster state and not warn anything if we let the next master finalizes the snapshot.

Makes sense :) I pushed 48a6eba

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I might be missing something, but this fix uses `unwrap()` which acts differently from `unwrapCause()` and does not rely on ElasticsearchWrapperException.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: `org.elasticsearch.ExceptionsHelper#unwrap` just loops through the `getCause` returns, it doesn't need `WrapperException` like `org.elasticsearch.ExceptionsHelper#unwrapCause`?

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: lol, thanks for the clarification. What a confusing bunch of methods :D

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Tanguy & Yannick!

**Pull request merged**